### PR TITLE
fix: "search-all-list" api endpoint and api tests 

### PR DESCRIPTION
**Pull request by bilaltahir21**
Jira: [ENT-5898](https://2u-internal.atlassian.net/jira/software/c/projects/ENT/boards/629?modal=detail&selectedIssue=ENT-5898)

### Bug
The `discovery.edx.org/api/v1/search/all/` endpoint returns unexpected results when the `exclude_expired_course_run` filter is applied in combination with other filters. It is because the `exclude_expired_course_run` is applied in the serializer after the results are filtered based on other filters like `published`.

### Expected Behaviour
search/all endpoint should give results after combining `exclude_expired_course_run` filter with other applied filters instead of just removing the expired course runs from the final filtered result. 

### Implementation Details
This PR is changing how the expired course runs are excluded from the search/all endpoint results. It excludes expired course runs based on the `exclude_expired_course_run` filter but includes the course runs which should be included due to the other filters. It improves the results returned when the `exclude_expired_course_run` filter is applied. This PR also included the tests for this scenario.

### CODE REVIEW

**Comment by openedx-webhooks** [comment]
Thanks for the pull request, @bilaltahir21!  Please note that it may take us up to several weeks or months to complete a review and merge your PR.

Feel free to add as much of the following information to the ticket as you can:
- supporting documentation
- [Open edX discussion forum threads](https://discuss.openedx.org/)
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).


Please let us know once your PR is ready for our review and all tests are green.


:warning: We can't start reviewing your pull request until you've submitted a     [signed contributor agreement](https://openedx.org/cla)     or indicated your institutional affiliation. Please see the [CONTRIBUTING](https://github.com/edx/edx-platform/blob/master/CONTRIBUTING.rst)     file for more information.     If you've signed an agreement in the past, you may need to re-sign.  See     [The New Home of the Open edX Codebase](https://open.edx.org/blog/the-new-home-of-the-open-edx-codebase/)     for details.

File: course_discovery/apps/course_metadata/search_indexes/serializers/course.py [hunk 1]
```diff
@@ -66,14 +66,26 @@ def get_course_runs(self, result):
         now = datetime.datetime.now(pytz.UTC)
         exclude_expired = request.GET.get('exclude_expired_course_run')
         detail_fields = request.GET.get('detail_fields')
+
+        ignored_params = []
```
Comment (HammadAhmadWaqas) [inline course_discovery/apps/course_metadata/search_indexes/serializers/course.py]: As we are using it anymore, Can we remove this variable?

**Comment by openedx-webhooks** [comment]
@bilaltahir21 Even though your pull request wasn’t merged, please take a moment  to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=openedx/course-discovery&entry.867055334=https://github.com/openedx/course-discovery/pull/3518&entry.1484655318=https://github.com/bilaltahir21&entry.752974735=2022-07-07+12:08&entry.1917517419=2022-07-13+11:06&entry.2133058324=No) so we can improve your experience in the future.

File: course_discovery/apps/course_metadata/search_indexes/serializers/course.py [hunk 2]
```diff
@@ -66,14 +66,26 @@ def get_course_runs(self, result):
...
+
         return [
             self.course_run_detail(request, detail_fields, course_run)
             for course_run in course_runs
             # Check if exclude_expire_course_run is in query_params then exclude the course
             # runs whose end date is passed. We do this here, rather than as an additional
             # `.exclude` because the course_runs have been prefetched by the read_queryset
             # of the search index.
-            if (not exclude_expired or course_run.end is None or course_run.end > now)
+            if should_include_course_run(course_run, dict(request.GET), exclude_expired)
```
Comment (HammadAhmadWaqas) [inline course_discovery/apps/course_metadata/search_indexes/serializers/course.py]: Query params are being sent with post data. so `request.GET` might not work here.

**Review by HammadAhmadWaqas — CHANGES_REQUESTED** [review]

File: course_discovery/apps/course_metadata/search_indexes/serializers/course.py (hunk 1 above)
Comment (bilaltahir21) [reply course_discovery/apps/course_metadata/search_indexes/serializers/course.py]: Removed.

File: course_discovery/apps/course_metadata/search_indexes/serializers/course.py (hunk 2 above)
Comment (bilaltahir21) [reply course_discovery/apps/course_metadata/search_indexes/serializers/course.py]: Replaced it with "query_params" and added another test for post request.

File: course_discovery/apps/api/v1/tests/test_views/test_search.py [hunk 3]
```diff
@@ -496,6 +520,111 @@ def test_query_count_exclude_expired_course_run(self, exclude_expired, expected_
...
+
+        response_data = response.json()
+
+        expected = [
+            self.course_run_detail(run)
+            for run in (course_run1, course_run2, course_run3)
+        ]
+        assert response_data['results'][0]['course_runs'] == expected
+
+    def test_results_filtered_by_exclude_expired_course_run_post(self):
```
Comment (HammadAhmadWaqas) [inline course_discovery/apps/api/v1/tests/test_views/test_search.py]: Can we use `ddt` here?

Comment (bilaltahir21) [reply course_discovery/apps/api/v1/tests/test_views/test_search.py]: Used DDT and changed serializer too.

File: course_discovery/apps/api/v1/tests/test_views/test_search.py [hunk 4]
```diff
@@ -496,6 +496,71 @@ def test_query_count_exclude_expired_course_run(self, exclude_expired, expected_
...
+        course.canonical_course_run = course_run1
+        course.save()
+
+        # Craeting seats to make course runs marketable
+        SeatFactory(course_run=course_run1)
+        SeatFactory(course_run=course_run2)
+        SeatFactory(course_run=course_run3)
+        SeatFactory(course_run=course_run4)
+
+        response = None
```
Comment (HammadAhmadWaqas) [inline course_discovery/apps/api/v1/tests/test_views/test_search.py:536]: Why we need to define response=None here?

**Review by HammadAhmadWaqas — APPROVED** [review]

File: course_discovery/apps/api/v1/tests/test_views/test_search.py (hunk 4 above)
Comment (bilaltahir21) [reply course_discovery/apps/api/v1/tests/test_views/test_search.py:536]: It is because we need `response` parameter outside the scope of `if` statement.

**Comment by openedx-webhooks** [comment]
@bilaltahir21 Even though your pull request wasn’t merged, please take a moment  to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=openedx/course-discovery&entry.867055334=https://github.com/openedx/course-discovery/pull/3518&entry.1484655318=https://github.com/bilaltahir21&entry.752974735=2022-07-07+12:08&entry.1917517419=2022-07-19+09:20&entry.2133058324=No) so we can improve your experience in the future.

**Comment by natabene** [comment]
@bilaltahir21 I am working with HelpDesk to get you added to the list. Sorry for the inconvenience.

**Comment by natabene** [comment]
@bilaltahir21 You CLA is all set. Now please edit the title and save (like add a space) and this will trigger the check to re-run and it will turn green. Can you please also remove the open-source-contribution label for me?

**Comment by bilaltahir21** [comment]
@natabene thank you very much 🙂 And I don't have access to add or remove labels from the pull request.

**Review by sameenfatima78 — APPROVED** [review]

**Review by uzairr — APPROVED** [review]
please add the implementation details with context in PR description. 👍

**Pull request merged**